src/theory/ff/sub_theory.h

```diff
@@ -0,0 +1,168 @@
+ *
+ * A field-specific theory
+ */
+
+#include "cvc5_private.h"
+
+#ifdef CVC5_USE_COCOA
+
+#ifndef CVC5__THEORY__FF__SUB_THEORY_H
+#define CVC5__THEORY__FF__SUB_THEORY_H
+
+#include <CoCoA/RingFp.H>
+
+#include <optional>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "context/cdlist_forward.h"
+#include "context/cdo.h"
+#include "context/context.h"
+#include "expr/node.h"
+#include "options/ff_options.h"
+#include "smt/env_obj.h"
+#include "theory/ff/core.h"
+#include "theory/ff/groebner.h"
+#include "theory/ff/stats.h"
+#include "theory/theory.h"
+#include "util/integer.h"
+
+namespace cvc5::internal {
+namespace theory {
+namespace ff {
+
+/**
+ * A solver for a specific, known finite field.
+ *
+ * While the main ff solver is responsible for all elements in any finite field,
+ * this solver just considers a single field. The main ff solver essentially
+ * just multiplexes between different sub-solvers.
+ *
+ * The solver only exposes a subset of the standard SMT theory interface. See
+ * the methods below.
+ *
+ * For now, our implementation assumes that the finite field has prime order.
+ */
+class SubTheory : protected EnvObj, protected context::ContextNotifyObj
+{
+ public:
+  // Create a new sub-theory.
+  //
+  // Parameters:
+  // * modulus: the size of this field for this theory, a prime.
+  SubTheory(Env& env, FfStatistics* stats, Integer modulus);
+
+  // Notify this theory of a node it may need to handle.
+  //
+  // All nodes this theory sees in the future must be pre-registered.
+  void preRegisterTerm(TNode);
+
+  // Assert a fact to this theory.
+  void notifyFact(TNode fact);
+
+  // Check the current facts.
+  void postCheck(Theory::Effort);
+
+  // Has a conflict been detected?
+  bool inConflict() const;
+
+  // What is that conflict?
+  const std::vector<Node>& conflict() const;
+
+  // Get a model.
+  const std::unordered_map<Node, Node>& model() const;
+
+ private:
+  // Called on SAT pop; we pop the incremental ideal if needed.
+  void contextNotifyPop() override;
+
+  // Compute a Groebner basis for the facts up to (not including) this index.
+  void computeBasis(size_t factIndex);
+
+  void extractModel();
+
+  // Initialize the polynomial ring, set up post-registration data structures.
+  void ensureInitPolyRing();
+
+  // Uninitialize the polynomial ring, clear post-registration data structures.
+  void clearPolyRing();
+
+  // Translate t to CoCoA, and cache the result.
+  void translate(TNode t);
+
+  // Is registration done and the polynomial ring initialized?
+  bool d_registrationDone();
+
+  // Manages the incremental GB.
+  std::optional<IncrementalIdeal> d_incrementalIdeal{};
+
+  // For an atom x == y, contains the potential inverse of (x - y).
+  //
+  // Used to make x != y.
+  std::unordered_map<Node, CoCoA::RingElem> d_atomInverses{};
+
+  // Facts, in notification order.
+  //
+  // Uses SAT context.
+  context::CDList<Node> d_facts;
```

Comment:
Its a bit funny to maintain this when the Theory also can reproduce the fact queue (from this member https://github.com/cvc5/cvc5/blob/main/src/theory/theory.h#L813).
Is it critical to maintain this list? Or could this be extracted from the fact queue from TheoryFf?

If you are interested in a standard way of accessing this list, I can give some guidance, e.g. using interfaces from the TheoryState utility (I'm somewhat surprised the class introduced in this PR does not depend on this utility already).

Reply:
This list contains only the facts for *this* sub-theory. That is, those in the specific field that this sub-theory manages.

We could extract those from the theory-wide list as needed, but keeping them here seemed asymptotically better?

We could also keep a list here of *indices* into the global list, which maybe fixes the asymptotic issue, but I'm not sure the global list is meant to support random access. Also, I don't really like the idea of keeping indices into a list that is very far away codebase-wise.

Thoughts?

Reply:
If we stick with this solution, I'll leave a comment about this conversation.

Reply:
Ok, this makes sense to me, yes comment in the doc on this.